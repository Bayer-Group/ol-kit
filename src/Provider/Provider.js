import React from 'react'
import PropTypes from 'prop-types'
import en from 'locales/en'
import ugh from 'ugh'

// context is only created when <Provider> is implemented (see constructor)
export let ProviderContext = null

/** A higher order component that provides context to connectToContext wrapped children
 * @component
 * @category Provider
 * @example ./example.md
 * @since 0.20.0
 */
class Provider extends React.Component {
  constructor (props) {
    super(props)

    // state becomes an object of persistedStateKeys (or component names) with their persisted states'
    this.state = {
      mapContext: {}
    }

    // create context when <Provider> is included in component tree
    ProviderContext = React.createContext()
  }

  persistState = (persistedState, persistedStateKey) => {
    if (!persistedStateKey || typeof persistedStateKey !== 'string') return ugh.error('persistedStateKey (string; usually "ComponentName") must be passed as second arg to persistState(state, persistedStateKey)') // eslint-disable-line no-console

    this.setState({ [persistedStateKey]: persistedState })
  }

  addMapToContext = mapContext => {
    this.setState({ mapContext })
  }

  getContextValue = () => {
    const { contextProps, map: mapProp, maps: mapsProp, translations } = this.props

    // if (!mapProp && !mapsProp.length) return ugh.throw('Provider requires either a \'map\' or \'maps\' prop to work!')

    // support multi-map components
    // default map reference to zeroith map from mapsProp if an array of maps has been passed
    const map = mapsProp.length ? mapsProp[0] : mapProp
    // default an array with single value of map prop if maps array has not been passed
    const maps = mapsProp.length ? mapsProp : [mapProp]

    return {
      addMapToContext: this.addMapToContext,
      map,
      maps,
      persistedState: this.state,
      persistState: this.persistState,
      translations,
      ...this.state.mapContext,
      ...contextProps
    }
  }

  render () {
    return (
      <ProviderContext.Provider value={this.getContextValue()}>
        {this.props.children}
      </ProviderContext.Provider>
    )
  }
}

Provider.defaultProps = {
  contextProps: {},
  maps: [],
  translations: en
}

Provider.propTypes = {
  /** Pass components as children of Provider component */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  /** Add any custom props to context and they will be passed to all components wrapped by connectToContext */
  contextProps: PropTypes.object,
  /** OL map object (required if maps array is not passed) */
  map: PropTypes.object,
  /** Array of Openlayers map objects for components that support multi-map (this will override anything passed to map prop) */
  maps: PropTypes.array,
  /** Object with key/value pairs for component translation strings */
  translations: PropTypes.object
}

export default Provider
