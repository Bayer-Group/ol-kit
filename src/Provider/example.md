The Provider component should wrap your entire application and never be unmounted. It will automatically interface with any ol-kit children within the component tree and connect them to context for things like `map`, `translations` and `selectInteraction` props. A `<Map>` mounts a `<Provider>` for you behind the scenes, but if you mount `<Provider>` declaratively (see below), components do not have to be children of `<Map>` (only `<Provider>`).

Fully integrated example:
```javascript static
import React from 'react'
import { Map, Popup, Provider } from '@bayer/ol-kit'

function App() {
  return (
    <Provider>
      <Map />
      <Popup />
    </Provider>
  )
}

export default App
```
Make sure the <Provider> never unmounts!
