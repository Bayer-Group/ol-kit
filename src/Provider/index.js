export { default as Provider, ProviderContext } from './Provider'
export * from './utils'
