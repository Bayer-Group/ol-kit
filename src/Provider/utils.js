import React from 'react'
import ugh from 'ugh'
import { ProviderContext } from 'Provider'

/**
 * A wrapper utility function designed to automatically pass down provider conntext as props from the Provider component
 * @function
 * @category Provider
 * @since 1.0.0
 * @param {Component} component - A React component you want wrapped
 * @returns {Component} A wrapped React component which will automatically be passed a reference to provider context
 */
export function connectToContext (Component) {
  if (!Component) return ugh.throw('Pass a React component to \'connectToContext\'')

  return props => { // eslint-disable-line react/display-name
    return !ProviderContext
      ? <Component {...props} />
      : (
        <ProviderContext.Consumer>
          {
            (providerProps = {}) => {
              // if propTypes is not defined on the component just pass all providerProps
              const filteredProviderProps = { ...providerProps }
              const { propTypes } = Component

              if (propTypes) {
                // filter out any props that do not need to get passed to this wrapped component
                Object.keys(providerProps).forEach(key => {
                  if (!propTypes[key]) delete filteredProviderProps[key]
                })
              }

              // // persistedState logic
              // const { persistedState, persistState } = providerProps
              // // set the key to look up component's persisted state
              // const persistedStateKey = props.persistedStateKey || Component.displayName || Component.name // eslint-disable-line react/prop-types

              return (
                <Component
                  // persistedState={persistedState[persistedStateKey]} // note: persistedState is undefined if persistedStateKey key doesn't exist yet (components should check for this)
                  // persistState={persistState}
                  {...filteredProviderProps}
                  {...props} />
              )
            }
          }
        </ProviderContext.Consumer>
      )
  }
}
