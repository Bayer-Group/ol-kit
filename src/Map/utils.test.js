import React from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import { waitFor } from '@testing-library/react'
import olInteractionSelect from 'ol/interaction/Select'
import olVectorLayer from 'ol/layer/Vector'
import qs from 'qs'
import {
  Map,
  addSelectInteraction,
  createMap,
  getSelectInteraction,
  updateMapFromUrl
} from 'Map'
import { connectToContext } from 'Provider'

describe('createMap', () => {
  // jest does not reset the DOM after each test, so we do this manually
  afterEach(() => {
    document.getElementsByTagName('html')[0].innerHTML = ''
  })

  it('createMap should return a map when given a DOM id', () => {
    const map = createMap({ target: 'test-id' })

    expect(typeof map).toEqual('object')
    expect(map.constructor.name).toBe('Map')
  })

  it('createMap should return a map when given a DOM element', () => {
    global.document.body.innerHTML = '<div id="map"></div>'
    const el = global.document.getElementById('map')

    const map = createMap({ target: el })

    expect(typeof map).toEqual('object')
    expect(map.constructor.name).toBe('Map')
  })

  it('createMap called without arguments throws an error', () => {
    const testCreate = () => createMap()

    expect(testCreate).toThrowError(new Error('You must pass an options object with a DOM target for the map'))
  })

  it('createMap called with incorrect arguments throws an error', () => {
    const testCreate = () => createMap({ target: true })

    expect(testCreate).toThrowError(new Error('The target should either by a string id of an existing DOM element or the element itself'))
  })
})

describe('connectToContext', () => {
  it('should render without passing a map', () => {
    // Map has not been mounted; no MapContext, so just render the Child
    const Consumer = connectToContext(props => <div>child comp</div>)
    const wrapper = mount(<Consumer inlineProp={true} />)

    // make sure connectToContext is passing inline props down to children
    expect(wrapper.props().inlineProp).toBe(true)
    // connectToContext should NOT add a map prop since Map is NOT mounted
    expect(wrapper.props().map).toBeUndefined()
  })

  it('should pass a map prop to children', async () => {
    const Child = props => <div>child comp</div>
    const Consumer = connectToContext(Child)
    const onMapInit = jest.fn()
    const wrapper = mount(<Map onMapInit={onMapInit}><Consumer inlineProp={true} /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())
    wrapper.update()

    expect(wrapper.find(Consumer).props().inlineProp).toBe(true)
    // make sure connectToContext is passing inline props down to children
    expect(wrapper.find(Child).props().inlineProp).toBe(true)
    // connectToContext should add map, selectInteraction, translations props since Map is mounted
    expect(wrapper.find(Child).props().map).toBeTruthy()
    expect(wrapper.find(Child).props().selectInteraction).toBeTruthy()
    expect(wrapper.find(Child).props().translations).toBeTruthy()
  })

  it('should filter out unneeded providerProps', async () => {
    const Child = props => <div>child comp</div>

    // this is defined to make sure uneeded propTypes get filtered out
    Child.propTypes = {
      inlineProp: PropTypes.bool,
      map: PropTypes.object
    }
    const Consumer = connectToContext(Child)
    const onMapInit = jest.fn()
    const wrapper = mount(<Map onMapInit={onMapInit}><Consumer inlineProp={true} /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())
    wrapper.update()

    expect(wrapper.find(Consumer).props().inlineProp).toBe(true)
    // make sure connectToContext is passing inline props down to children
    expect(wrapper.find(Child).props().inlineProp).toBe(true)
    // connectToContext should add a map prop since Map is mounted and defined in propTypes
    expect(wrapper.find(Child).props().map).toBeTruthy()
    // connectToContext should not pass extra provided props down since propTypes do not include these props:
    expect(wrapper.find(Child).props().translations).toBeUndefined()
    expect(wrapper.find(Child).props().selectInteraction).toBeUndefined()
  })
})

describe('updateMapFromUrl', () => {
  // jest does not reset the DOM after each test, so we do this manually
  afterEach(() => {
    document.getElementsByTagName('html')[0].innerHTML = ''
  })

  it('should throw error for missing map arg', () => {
    expect(updateMapFromUrl()).rejects.toMatchSnapshot()
  })

  it('should return resolved promise for missing view url param', () => {
    global.document.body.innerHTML = '<div id="map"></div>'

    const el = global.document.getElementById('map')
    const map = createMap({ target: el })

    expect(updateMapFromUrl(map)).resolves.toMatchSnapshot()
  })

  it('should update map from url with default "view" url param', () => {
    global.document.body.innerHTML = '<div id="map"></div>'

    const el = global.document.getElementById('map')
    const map = createMap({ target: el })

    // set initial center, zoom & rotation
    map.getView().setCenter([0, 0])
    map.getView().setZoom(3)
    map.getView().setRotation(0)

    expect(map.getView().getCenter()).toEqual([0, 0])
    expect(map.getView().getZoom()).toBe(3)
    expect(map.getView().getRotation()).toBe(0)

    // set the url with a view param
    window.history.replaceState(null, '', `${window.location.pathname}?view=49.618551,-97.280674,8.00,0.91`)

    // second arg (viewParam) is defaulted to "view"
    updateMapFromUrl(map)
    // updated map center, zoom & rotation from url
    expect(map.getView().getCenter()).toEqual([-10829235.09370645, 6380475.798452517])
    expect(map.getView().getZoom()).toBe(8)
    // round off crazy long decimal
    expect(Number(map.getView().getRotation().toFixed(2))).toEqual(0.91)
  })

  it('should update map from url with custom "customParam" url param', () => {
    global.document.body.innerHTML = '<div id="map"></div>'

    const el = global.document.getElementById('map')
    const map = createMap({ target: el })

    // set initial center, zoom & rotation
    map.getView().setCenter([0, 0])
    map.getView().setZoom(3)
    map.getView().setRotation(0)

    expect(map.getView().getCenter()).toEqual([0, 0])
    expect(map.getView().getZoom()).toBe(3)
    expect(map.getView().getRotation()).toBe(0)

    // set the url with a view param
    window.history.replaceState(null, '', `${window.location.pathname}?customParam=49.618551,-97.280674,8.00,0.91`)

    // second arg (viewParam) is "customParam"
    updateMapFromUrl(map, 'customParam')
    // updated map center, zoom & rotation from url
    expect(map.getView().getCenter()).toEqual([-10829235.09370645, 6380475.798452517])
    expect(map.getView().getZoom()).toBe(8)
    // round off crazy long decimal
    expect(Number(map.getView().getRotation().toFixed(2))).toEqual(0.91)
  })
})

describe('updateUrlFromMap', () => {
  // jest does not reset the DOM after each test, so we do this manually
  afterEach(() => {
    document.getElementsByTagName('html')[0].innerHTML = ''
  })

  it('should update url from map changes', async () => {
    global.document.body.innerHTML = '<div id="map"></div>'
    // set the url with a competing url param
    window.history.replaceState(null, '', `${window.location.pathname}?existingParam=true&otherParam=false`)

    const onMapInit = async map => {
      const query = qs.parse(window.location.search, { ignoreQueryPrefix: true })

      // existingParam is set above when the url is reset^ (make sure it still exists)
      expect(query.existingParam).toBe('true')
      // check to make sure the param otherParam set to the url hasn't been overwritten
      expect(query.otherParam).toBe('false')
      // waitFor allows the map to load/moveend which triggers the url update before checking the query to see if the param exists
      // check to make sure the view param was added to the url
      await waitFor(() => expect(query.view).toBe('39.000000,-96.000000,5.00,0.00'))
    }

    // default updateUrlFromView is true (which is what we're testing here)
    mount(<Map onMapInit={onMapInit} />)
  })
})

describe('selectInteractions', () => {
  // jest does not reset the DOM after each test, so we do this manually
  afterEach(() => {
    document.getElementsByTagName('html')[0].innerHTML = ''
  })

  it('should create and find default select interaction', async () => {
    global.document.body.innerHTML = '<div id="map"></div>'

    const onMapInit = async map => {
      const defaultSelect = getSelectInteraction(map)

      // find default select without providing a name arg
      expect(defaultSelect).toBeInstanceOf(olInteractionSelect)
      const selectInteractionsOnMap = map.getInteractions().getArray().filter(interaction => {
        return interaction instanceof olInteractionSelect
      })

      expect(selectInteractionsOnMap.length).toBe(1)

      const { layer, select } = addSelectInteraction(map, 'custom_test_select')

      // test addSelectInteraction
      expect(select).toBeInstanceOf(olInteractionSelect)
      expect(layer).toBeInstanceOf(olVectorLayer)
      // new select should be on the map
      expect(getSelectInteraction(map, 'custom_test_select')).toBeInstanceOf(olInteractionSelect)
      const selectInteractionsOnMapAgain = map.getInteractions().getArray().filter(interaction => {
        return interaction instanceof olInteractionSelect
      })

      expect(selectInteractionsOnMapAgain.length).toBe(2)
    }

    // default updateUrlFromView is true (which is what we're testing here)
    mount(<Map onMapInit={onMapInit} />)
  })

  it('should throw an error getting a select interaction by bad name', async () => {
    global.document.body.innerHTML = '<div id="map"></div>'

    const onMapInit = async map => {
      // this named select has never been added to the map- should throw error
      expect(getSelectInteraction(map, 'bad_bad_fake_name')).toThrowErrorMatchingSnapshot()
    }

    // default updateUrlFromView is true (which is what we're testing here)
    mount(<Map onMapInit={onMapInit} />)
  })
})
