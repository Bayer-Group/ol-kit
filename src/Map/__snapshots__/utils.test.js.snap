// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateMapFromUrl should return resolved promise for missing view url param 1`] = `undefined`;

exports[`updateMapFromUrl should throw error for missing map arg 1`] = `"'updateMapFromUrl' requires a valid openlayers map as the first argument"`;
